---
title: "Clustering"
output: html_document
date: "2025-04-24"
params:
  input_rds: NULL
  output_rds: NULL
  pca: 10
---


```{r}
#Importing libraries
library(dplyr)
library(Seurat)
library(patchwork)
```
Seurat applies a graph-based clustering approach.
As in PhenoGraph, we first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the FindNeighbors() function, and takes as input the previously defined dimensionality of the dataset (first 10 PCs).

To cluster the cells, we next apply modularity optimization techniques such as the Louvain algorithm (default) or SLM [SLM, Blondel et al., Journal of Statistical Mechanics], to iteratively group cells together, with the goal of optimizing the standard modularity function. The FindClusters() function implements this procedure, and contains a resolution parameter that sets the ‘granularity’ of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters can be found using the Idents() function.

```{r}
# Load input object
pbmc <- readRDS(params$input_rds)
pbmc <- FindNeighbors(pbmc, dims = 1:params$pca) # Aqui marquem la diemnsionalitat escollida (PC 10 en aquest cas)
pbmc <- FindClusters(pbmc, resolution = 0.5)

# Look at cluster IDs of the first 5 cells
head(Idents(pbmc), 5)
```
RUN NON-LINEAR DIEMNSIONAL REDUCTION (UMAP/tSNE)

he goal of these algorithms is to learn underlying structure in the dataset, in order to place similar cells together in low-dimensional space. Therefore, cells that are grouped together within graph-based clusters determined above should co-localize on these dimension reduction plots. 

While we and others have routinely found 2D visualization techniques like tSNE and UMAP to be valuable tools for exploring datasets, all visualization techniques have limitations, and cannot fully represent the complexity of the underlying data. In particular, these methods aim to preserve local distances in the dataset (i.e. ensuring that cells with very similar gene expression profiles co-localize), but often do not preserve more global relationships. We encourage users to leverage techniques like UMAP for visualization, but to avoid drawing biological conclusions solely on the basis of visualization techniques.

```{r}
pbmc <- RunUMAP(pbmc, dims = 1:params$pca)
```
```{r}
# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
DimPlot(pbmc, reduction = "umap")
```

en aquest punt es recomana guardar el objecte modificat per no haver de repetir els passos. 
```{r}
# Save object
saveRDS(pbmc, params$output_rds)
```